In [71]:
import pandas as pd
import numpy as np

# Loding dataset

In [51]:
dataset = pd.read_csv('StudentsPerformance.csv')
dataset

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

# Checking null values in the dataset

In [70]:
dataset.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
math var                       0
reading var                    0
writing var                    0
dtype: int64

## There are no null values in the dataset

# 1. Find out how many males and females participated in the test.

In [52]:
dataset['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

### 518 females and 482 males participated in the test. 

# 2. What do you think about the students' parental level of education?

In [53]:
dataset['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

### Most of the parents are educated and 59 of them has master's degree

# 3. a) Who scores the most on average for math, reading and writing based on Gender

In [54]:
ms = sp.groupby('gender')['math score'].mean()
ms

gender
female    63.376448
male      68.821577
Name: math score, dtype: float64

### Male students have an avereage math score compared to female students

In [55]:
rs = sp.groupby('gender')['reading score'].mean()
rs

gender
female    72.590734
male      65.545643
Name: reading score, dtype: float64

### Female students have an average reading score compared to male students

In [56]:
ws = sp.groupby('gender')['writing score'].mean()
ws

gender
female    72.467181
male      63.446058
Name: writing score, dtype: float64

### Female students have an average writing score compared to male students

# 3.b) Who scores the most on average for math, reading and writing based on Test preparation course

In [57]:
tp1 = sp.groupby('test preparation course')['math score'].mean()
tp1

test preparation course
completed    69.96648
none         63.78972
Name: math score, dtype: float64

In [58]:
tp2 = sp.groupby('test preparation course')['reading score'].mean()
tp2

test preparation course
completed    74.175978
none         66.417445
Name: reading score, dtype: float64

In [59]:
tp3 = sp.groupby('test preparation course')['writing score'].mean()
tp3

test preparation course
completed    74.684358
none         64.457944
Name: writing score, dtype: float64

### In all 3, average score is more for the student who completed the test preparation course

# 4.a)  The scoring variation for math, reading and writing based on Gender

In [60]:
dataset_var = dataset
dataset_var['math var'] = (dataset['math score']-dataset['math score'].mean())**2
dataset_var['reading var'] = (dataset_var['reading score']-dataset_var['reading score'].mean())**2
dataset_var['writing var'] = (dataset_var['writing score']-dataset_var['writing score'].mean())**2
dataset_var

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  \
0                      none           0             17             10   
1                      none           8             24             23   
2                      none          18             32             28   
3                      none          11             38             32   
4                      none          22             39             33   
..                      ...         ...            ...            ...   
995               completed          99             87             81   
996               completed         100             96             86   
997               completed         100             97             99   
998               completed         100            100             93   
999               completed         100            100            100   

        math var  reading var  writing var  
0    4356.132001  2724.318025  3377.818161  
1    3364.116001  2042.588025  2035.724161  
2    2304.096001  1383.468025  1609.534161  
3    3025.110001   973.128025  1304.582161  
4    1936.088001   911.738025  1233.344161  
..           ...          ...          ...  
995  1088.934001   317.018025   165.920161  
996  1155.932001   718.508025   319.730161  
997  1155.932001   773.118025   953.636161  
998  1155.932001   948.948025   619.064161  
999  1155.932001   948.948025  1016.398161  

[1000 rows x 11 columns]

## Scoring variation for reading based on gender

In [61]:
dataset.groupby('gender')['math var'].sum()/dataset['gender'].value_counts()

gender
female    263.350808
male      219.405146
dtype: float64

## Scoring variation for writing based on gender

In [62]:
dataset.groupby('gender')['reading var'].sum()/dataset['gender'].value_counts()

gender
female    218.807523
male      213.113440
dtype: float64

## Scoring variation for math based on gender

In [63]:
dataset.groupby('gender')['writing var'].sum()/dataset['gender'].value_counts()

gender
female    238.850586
male      223.830364
dtype: float64

### Females students have are high scoring variation for maths,reading and writing compared to males students

# 4.b) The scoring variation for math, reading and writing based on Test preparation course

## Scoring variation for math based on Test Preparation Course

In [64]:
dataset.groupby('test preparation course')['math var'].sum()/dataset['test preparation course'].value_counts()

completed    226.020001
none         251.174203
dtype: float64

## Scoring variation for reading based on Test Preparation Course

In [65]:
dataset.groupby('test preparation course')['reading var'].sum()/dataset['test preparation course'].value_counts()

completed    207.564086
none         220.802231
dtype: float64

## Scoring variation for writing based on Test Preparation Course

In [66]:
dataset.groupby('test preparation course')['writing var'].sum()/dataset['test preparation course'].value_counts()

completed    217.817139
none         239.302653
dtype: float64

### Scoring variation of students doesn't complete test preparation course is higher compared to the students completed the test preparation course in all 3 subjects

# 5. The management needs your help to give bonus points to the top 25% of students based on their maths score, so how will you help the management to achieve this

In [67]:

dataset['math score'].describe()

count    1000.000000
mean       66.001000
std        15.569567
min         0.000000
25%        57.000000
50%        66.000000
75%        77.000000
max       100.000000
Name: math score, dtype: float64

### 3rd quartile of maths score is 77 marks.75% of the students have marks less than 77, and 25% of students have 77 and above marks. Students eligible for the bonus points are marks 77 or above.

In [68]:
dataset[dataset['math score']>=77]

gender race/ethnicity parental level of education         lunch  \
414  female        group A                 high school  free/reduced   
415  female        group B             master's degree  free/reduced   
416  female        group B           bachelor's degree  free/reduced   
417  female        group B             master's degree      standard   
418  female        group B                 high school      standard   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  \
414               completed          77             88             85   
415               completed          77             97             94   
416                    none          77             85             87   
417                    none          77             90             84   
418               completed          77             82             89   
..                      ...         ...            ...            ...   
995               completed          99             87             81   
996               completed         100             96             86   
997               completed         100             97             99   
998               completed         100            100             93   
999               completed         100            100            100   

        math var  reading var  writing var  
414   120.978001   353.628025   284.968161  
415   120.978001   773.118025   669.826161  
416   120.978001   249.798025   356.492161  
417   120.978001   432.848025   252.206161  
418   120.978001   163.968025   436.016161  
..           ...          ...          ...  
995  1088.934001   317.018025   165.920161  
996  1155.932001   718.508025   319.730161  
997  1155.932001   773.118025   953.636161  
998  1155.932001   948.948025   619.064161  
999  1155.932001   948.948025  1016.398161  

[255 rows x 11 columns]

### 255 students score 77 marks and above.

In [69]:
dataset.sort_values('math score',ascending=False).head(250)


gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
856    male        group E            some high school      standard   
855    male        group E          associate's degree  free/reduced   
854    male        group D            some high school      standard   
853    male        group D          associate's degree  free/reduced   
852    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
999               completed         100            100            100   
996               completed         100             96             86   
515                    none         100             92             97   
517                    none         100            100            100   
516                    none         100            100            100   
..                      ...         ...            ...            ...   
856               completed          77             76             77   
855               completed          77             69             68   
854               completed          77             68             69   
853                    none          77             78             73   
852                    none          77             62             64   

        math var  reading var  writing var  
999  1155.932001   948.948025  1016.398161  
996  1155.932001   718.508025   319.730161  
515  1155.932001   520.068025   834.112161  
517  1155.932001   948.948025  1016.398161  
516  1155.932001   948.948025  1016.398161  
..           ...          ...          ...  
856   120.978001    46.308025    78.872161  
855   120.978001     0.038025     0.014161  
854   120.978001     1.428025     0.776161  
853   120.978001    77.528025    23.824161  
852   120.978001    51.768025    16.966161  

[250 rows x 11 columns]

### After sorting the math score column in descending order, top 250 students are eligible for bonus points and five of them lose their bonus points.